pkg/debug/target.go

```diff
@@ -0,0 +1,289 @@
+package debug
+
+import (
+	"errors"
+	"fmt"
+	"go/ast"
+	"runtime"
+
+	"github.com/go-delve/delve/pkg/proc"
+	"github.com/go-delve/delve/pkg/proc/core"
+	"github.com/go-delve/delve/pkg/proc/gdbserial"
+	"github.com/go-delve/delve/pkg/proc/native"
+)
+
+// Target represents the process being debugged.
+// It is responsible for implementing the high level logic
+// that is used to manipulate and inspect a running process.
+type Target struct {
+	proc.Process
+
+	bi *proc.BinaryInfo
+}
+
+// New returns an initialized Target.
+func New(p proc.Process, os, arch string, debugInfoDirs []string) (*Target, error) {
```

Comment:
I don't like this, Launch, Attach and OpenCoreOrRecording. I think the way we have it now, with the backends creating and returing a proc.Process object is better. I do like the debug.Target struct but I think we should keep the setup where initialization is done in the backends and they instead return a debug.Target object.

Doing it this way (with debug.Launch, debug.Attach...) leads to the awkward and confusing SetTarget method. I know this is just a step towards a refactoring where SetTarget shouldn't be necessary but (a) I don't think it's possible to rewrite the backends in such a way that they don't need to access the Target object, (b) even if it's possible I don't believe it to be a worthwhile goal, in fact I think it will make delve's codebase less flexible.

Doing it this way also makes the initialization procedure more complicated: note for example how the core backend needs to return the os/arch pair to pkg/debug which will then call an Initialization method inside the core package, vs. how it worked before.

Reply:
> I don't like this, Launch, Attach and OpenCoreOrRecording. I think the way we have it now, with the backends creating and returing a proc.Process object is better. I do like the debug.Target struct but I think we should keep the setup where initialization is done in the backends and they instead return a debug.Target object.
> 

I disagree; one large goal of this refactor is to reduce duplication and make the backends as simple as possible. They should, as much as possible, _only_ know how to do low-level process related tasks (continue a process, read memory, write memory, stop a process, etc) with a higher level package building on top of the simple interface. If we don't centralize the initialization logic then we have every single backend knowing how to initialize the BinaryInfo object, so we repeat that code with every backend. 

> Doing it this way (with debug.Launch, debug.Attach...) leads to the awkward and confusing SetTarget method. I know this is just a step towards a refactoring where SetTarget shouldn't be necessary but (a) I don't think it's possible to rewrite the backends in such a way that they don't need to access the Target object, (b) even if it's possible I don't believe it to be a worthwhile goal, in fact I think it will make delve's codebase less flexible.
> 

This is why I also wanted to post to full change list. This was just a refactor step which let me start breaking things up while still being able to run the tests locally and ensure the refactor wasn't breaking anything. It's temporary and is not present in the final commits of the larger PR.

> Doing it this way also makes the initialization procedure more complicated: note for example how the core backend needs to return the os/arch pair to pkg/debug which will then call an Initialization method inside the core package, vs. how it worked before.

One backend having to return a few more parameters which only the `debug` package sees seems like a win over _every_ backend having to repeat startup and initialization code.

Reply:
I've replaced `SetTarget` with `SetBinaryInfo` to simply re-inject it into the `proc.Process`. This is still a win overall because the actual creation of the BinaryInfo is still done at a higher level. Again, I think eventually this should not be necessary.

Reply:
>I disagree; one large goal of this refactor is to reduce duplication and make the backends as simple as possible

I think we have hit an irreconciliable difference of opinion. I don't think what you want to do can or should be done and I don't see myself ever being swayed to the contrary.

Let's approach this from a different angle. Let's call point A the state of the code base as it is now. And let's call B the state of the code base as it would be after this PR. Let's call C the state where debug.Target gets introduced but the initialization procedure of the backend does not change and there is no debug.Launch/debug.Attach/etc. This introduces debug.Target which we both like and gets us closer to removing proc.CommonProcess which we both don't like, right? Can we also agree that point C is intermediate between A and B? If so can we just get to C for now, finish removing CommonProcess and then revisit the initialization procedure later?

Reply:
I'm 100% fine with compromising on option C if only for the sake of moving forward, but I kind of want to understand your hesitation to the consolidation of the startup code. From my perspective, of all the proposed changes this seems the most benign.

Here's my thoughts and justification on this change:

1. Less code will always decrease maintenance load. If packageA, packageB and packageC (native, gdbserial, core) all must perform the same action, and as a result of that end up producing a `debug.Target` struct, why not move all the code that all 3 packages are copying from each other into one central location?
2. There will only ever be one, single `debug.Target` implementation. There are already 3 backend implementations and there's the possibility for more. The more logic we consolidate into `debug.Target` means less logic _every_ _single_ backend will have to implement and duplicate, which I would argue leads to fewer bugs in the long run. Right now both native and gdbserial know about the goroutine cache. What if we find a bug where the cache isn't being cleared properly? Now we have to make sure it's fixed 2 places instead of just one. How is that a win?
3. Having the backend packages know about `*debug.Target` seems backwards to me and provides a leaky abstraction. In this case, now everything in the project has to know about the various backend packages AND the `debug` package, so instead of simplifying things and removing we're now instead _adding_ a new concept _on_ _top_ of the existing concepts. This refactor is meant to be reductive not additive.

Other than the fact that it's going to be tricky and must be done with care, I'm still not convinced that simplifying the backends in any way we possibly can would hurt us in the long run. The simpler and dumber they are, the easier it will be to add new OS support. If all we have to do is satisfy a tiny interface that is basically "start process, stop process, read memory, write memory, set breakpoint (in the case of the gdbserial backend we can leave this for now)" then adding a new backend becomes trivial. It doesn't have to worry if it's clearing the G cache, or creating the binary info struct, or setting the "current thread" or "current goroutine". Everything that is common and _must_ be done by all backends should be lifted out and implemented in one single place. Then we only ever have to write that code once, and we can add a million backends after that, and that code will never be duplicated.

Reply:
1. I don't think this actually removes much duplicate code, anything that can be done with this can be done with PostInitializationSetup.

2. It makes the control flow more complicated. For example now a backend's Launch in the backend calls NewBreakpointMap, NewBinaryInfo and then PostInitializationSetup. With this PR debug.Launch calls backend.Launch which does some work, returns to debug.Launch which calls debug.(*Target).Initialize which calls back the backend's Initialize function. I don't like the control bounching back and forth between debug and the backend.

3. The final version of this PR has a couple of downcasts in debug.(*Target).Initialize to execute different code for special backends, that suggests this is the wrong approach.

4. I think nobody will ever write another implementation of proc.Process, ever. The last time it happened it was april 2017, roughly 60% of the way through the entire life of this project.

Reply:
I agree with the points you made, especially regarding having to type cast and having control flow bouncing back and forth. I'll rework this a bit.